bug: jpa persistence not stable under concurrent load

### Pre-flight

- [x] I’m using the latest release or a recent snapshot.
- [x] I searched existing issues and Discussions.

### What happened?

The JPA/PostgreSQL persistence mode is currently not stable under concurrent load. Starting around 60-100 requests per second, the server logs show duplicate key violations in TaskInfoEntity, even though the HTTP endpoint still returns success.

The same _jsonpointer_ is inserted more than once during `writeCompletedTask`, suggesting the same task completion is persisted concurrently or that _jsonpointer_ is not unique enough as a primary key.

### Minimal reproducer

```bash
1. Start hello-world flow
2. Launch simple benchmark: k6 run -e RATE=65 -e DURATION=1m bench/k6-hello.js
```

### Quarkus version

3.32.2

### quarkus-flow version

1.0.0-SNAPSHOT

### Java version

_No response_

### Runtime

JVM (dev/test)

### Quarkus extensions in use

[agroal, cdi, flow, flow-persistence-jpa, hibernate-orm, hibernate-orm-panache, jackson-jq, jdbc-postgresql, micrometer, narayana-jta, rest, rest-client, rest-client-jackson, rest-jackson, smallrye-context-propagation, vertx]

### Relevant logs/stack traces

```text
2026-03-09 20:49:58,367 WARN  [org.hibernate.orm.jdbc.error] (executor-thread-9) HHH000247: ErrorCode: 0, SQLState: 23505
2026-03-09 20:49:58,368 WARN  [org.hibernate.orm.jdbc.error] (executor-thread-9) ERROR: duplicate key value violates unique constraint "taskinfoentity_pkey"
  Detail: Key (jsonpointer)=(do/0/ac094b7a-ed7d-4614-94a8-0bfe60a08e34) already exists.
2026-03-09 20:49:58,368 WARN  [com.arjuna.ats.arjuna] (executor-thread-9) ARJUNA012125: TwoPhaseCoordinator.beforeCompletion - failed for SynchronizationImple< 0:ffff7f000001:9ff7:69af2462:113b, org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization@53a86da7 >: org.hibernate.exception.ConstraintViolationException: could not execute statement [ERROR: duplicate key value violates unique constraint "taskinfoentity_pkey"
  Detail: Key (jsonpointer)=(do/0/ac094b7a-ed7d-4614-94a8-0bfe60a08e34) already exists.] [insert into TaskInfoEntity (context,instant,isEndNode,model,nextPosition,task_type,jsonPointer) values (?,?,?,?,?,1,?)]
	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:73)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:34)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:115)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:184)
	at org.hibernate.engine.jdbc.mutation.internal.AbstractMutationExecutor.performNonBatchedMutation(AbstractMutationExecutor.java:145)
	at org.hibernate.engine.jdbc.mutation.internal.MutationExecutorSingleNonBatched.performNonBatchedOperations(MutationExecutorSingleNonBatched.java:53)
	at org.hibernate.engine.jdbc.mutation.internal.AbstractMutationExecutor.execute(AbstractMutationExecutor.java:66)
	at org.hibernate.engine.jdbc.mutation.internal.AbstractMutationExecutor.execute(AbstractMutationExecutor.java:55)
	at org.hibernate.persister.entity.mutation.InsertCoordinatorStandard.doStaticInserts(InsertCoordinatorStandard.java:180)
	at org.hibernate.persister.entity.mutation.InsertCoordinatorStandard.coordinateInsert(InsertCoordinatorStandard.java:120)
	at org.hibernate.persister.entity.mutation.InsertCoordinatorStandard.insert(InsertCoordinatorStandard.java:93)
	at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:108)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:634)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:505)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:381)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:138)
	at org.hibernate.internal.SessionImpl.fireFlush(SessionImpl.java:1484)
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:481)
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:2111)
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2033)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:410)
	at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl.beforeCompletion(JtaTransactionCoordinatorImpl.java:323)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl.beforeCompletion(SynchronizationCallbackCoordinatorNonTrackingImpl.java:38)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization.beforeCompletion(RegisteredSynchronization.java:32)
	at com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple.beforeCompletion(SynchronizationImple.java:52)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.beforeCompletion(TwoPhaseCoordinator.java:348)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.end(TwoPhaseCoordinator.java:66)
	at com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:135)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1315)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:104)
	at io.quarkus.narayana.jta.runtime.NotifyingTransactionManager.commit(NotifyingTransactionManager.java:70)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.endTransaction(TransactionalInterceptorBase.java:406)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:175)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:107)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.doIntercept(TransactionalInterceptorRequired.java:38)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.intercept(TransactionalInterceptorBase.java:61)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.intercept(TransactionalInterceptorRequired.java:32)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)
	at io.quarkiverse.flow.persistence.jpa.JpaInstanceOperations_Subclass.writeCompletedTask(Unknown Source)
	at io.quarkiverse.flow.persistence.jpa.JpaInstanceOperations_ClientProxy.writeCompletedTask(Unknown Source)
	at io.serverlessworkflow.impl.persistence.AbstractPersistenceInstanceWriter.lambda$taskCompleted$3(AbstractPersistenceInstanceWriter.java:65)
	at io.quarkiverse.flow.persistence.jpa.JpaInstanceWriter.lambda$doTransaction$0(JpaInstanceWriter.java:37)
	at java.base/java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:787)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
	at io.smallrye.context.impl.wrappers.SlowContextualRunnable.run(SlowContextualRunnable.java:19)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:677)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "taskinfoentity_pkey"
  Detail: Key (jsonpointer)=(do/0/ac094b7a-ed7d-4614-94a8-0bfe60a08e34) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2875)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2560)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:429)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:526)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:436)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:196)
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:157)
	at io.agroal.pool.wrapper.PreparedStatementWrapper.executeUpdate(PreparedStatementWrapper.java:90)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:181)
	... 54 more
```

--- a/persistence/jpa/runtime/src/main/java/io/quarkiverse/flow/persistence/jpa/CompletedTaskEntity.java
+++ b/persistence/jpa/runtime/src/main/java/io/quarkiverse/flow/persistence/jpa/CompletedTaskEntity.java
@@ -26,10 +26,10 @@ public class CompletedTaskEntity extends TaskInfoEntity {
     public CompletedTaskEntity() {
     }
 
-    public CompletedTaskEntity(String jsonPointer, Instant instant, WorkflowModel model, WorkflowModel context,
+    public CompletedTaskEntity(TaskInfoKey key, Instant instant, WorkflowModel model, WorkflowModel context,
             boolean isEndNode,
             String nextPosition) {
-        super(jsonPointer);
+        super(key);
         this.instant = instant;
         this.model = model;
         this.context = context;

--- a/persistence/jpa/runtime/src/main/java/io/quarkiverse/flow/persistence/jpa/JpaInstanceOperations.java
+++ b/persistence/jpa/runtime/src/main/java/io/quarkiverse/flow/persistence/jpa/JpaInstanceOperations.java
@@ -8,6 +8,7 @@ import java.util.stream.Stream;
 
 import jakarta.enterprise.context.ApplicationScoped;
 import jakarta.inject.Inject;
+import jakarta.persistence.EntityManager;
 import jakarta.transaction.Transactional;
 
 import io.serverlessworkflow.impl.TaskContext;
@@ -31,6 +32,9 @@ public class JpaInstanceOperations implements PersistenceInstanceOperations {
     @Inject
     ProcessInstanceRepository repository;
 
+    @Inject
+    EntityManager em;
+
     @Override
     public void writeInstanceData(WorkflowContextData workflowContext) {
         WorkflowInstanceData instance = workflowContext.instanceData();
@@ -40,18 +44,19 @@ public class JpaInstanceOperations implements PersistenceInstanceOperations {
 
     @Override
     public void writeRetryTask(WorkflowContextData workflowContext, TaskContextData taskContext) {
-        find(workflowContext).getTasks()
-                .add(new RetriedTaskEntity(taskContext.position().jsonPointer(), ((TaskContext) taskContext).retryAttempt()));
+        em.persist(new RetriedTaskEntity(TaskInfoKey.from(workflowContext, taskContext),
+                ((TaskContext) taskContext).retryAttempt()));
     }
 
     @Override
     public void writeCompletedTask(WorkflowContextData workflowContext, TaskContextData taskContext) {
         TransitionInfo transition = ((TaskContext) taskContext).transition();
         AbstractTaskExecutor<?> next = (AbstractTaskExecutor<?>) transition.next();
-        find(workflowContext).getTasks().add(new CompletedTaskEntity(
-                taskContext.position().jsonPointer(), taskContext.completedAt(), taskContext.output(),
-                workflowContext.context(),
-                transition.isEndNode(), next == null ? null : next.position().jsonPointer()));
+        em.persist(
+                new CompletedTaskEntity(
+                        TaskInfoKey.from(workflowContext, taskContext), taskContext.completedAt(), taskContext.output(),
+                        workflowContext.context(),
+                        transition.isEndNode(), next == null ? null : next.position().jsonPointer()));
 
     }
 

--- a/persistence/jpa/runtime/src/main/java/io/quarkiverse/flow/persistence/jpa/ProcessInstanceEntity.java
+++ b/persistence/jpa/runtime/src/main/java/io/quarkiverse/flow/persistence/jpa/ProcessInstanceEntity.java
@@ -5,12 +5,11 @@ import java.util.Collection;
 
 import jakarta.persistence.Basic;
 import jakarta.persistence.CascadeType;
+import jakarta.persistence.Column;
 import jakarta.persistence.Entity;
 import jakarta.persistence.FetchType;
 import jakarta.persistence.Id;
 import jakarta.persistence.IdClass;
-import jakarta.persistence.JoinColumn;
-import jakarta.persistence.JoinColumns;
 import jakarta.persistence.OneToMany;
 
 import org.hibernate.annotations.DynamicUpdate;
@@ -29,17 +28,16 @@ public class ProcessInstanceEntity {
     @Id
     private String applicationId;
 
-    @Basic(fetch = FetchType.LAZY, optional = true)
+    @Column(nullable = true)
     private WorkflowStatus status;
 
-    @Basic(fetch = FetchType.LAZY)
+    @Column(nullable = false)
     private Instant startedAt;
 
     @Basic(fetch = FetchType.LAZY)
     private WorkflowModel input;
 
-    @OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
-    @JoinColumns({ @JoinColumn(name = "processInstanceId"), @JoinColumn(name = "applicationId") })
+    @OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL, mappedBy = "processInstance")
     private Collection<TaskInfoEntity> tasks;
 
     public ProcessInstanceEntity() {

--- a/persistence/jpa/runtime/src/main/java/io/quarkiverse/flow/persistence/jpa/RetriedTaskEntity.java
+++ b/persistence/jpa/runtime/src/main/java/io/quarkiverse/flow/persistence/jpa/RetriedTaskEntity.java
@@ -14,8 +14,8 @@ public class RetriedTaskEntity extends TaskInfoEntity {
     @Column
     private short retryAttempt;
 
-    public RetriedTaskEntity(String jsonPointer, short retryAttempt) {
-        super(jsonPointer);
+    public RetriedTaskEntity(TaskInfoKey key, short retryAttempt) {
+        super(key);
         this.retryAttempt = retryAttempt;
     }
 

--- a/persistence/jpa/runtime/src/main/java/io/quarkiverse/flow/persistence/jpa/TaskInfoEntity.java
+++ b/persistence/jpa/runtime/src/main/java/io/quarkiverse/flow/persistence/jpa/TaskInfoEntity.java
@@ -2,30 +2,36 @@ package io.quarkiverse.flow.persistence.jpa;
 
 import jakarta.persistence.DiscriminatorColumn;
 import jakarta.persistence.DiscriminatorType;
+import jakarta.persistence.EmbeddedId;
 import jakarta.persistence.Entity;
-import jakarta.persistence.GeneratedValue;
-import jakarta.persistence.Id;
+import jakarta.persistence.FetchType;
 import jakarta.persistence.Inheritance;
 import jakarta.persistence.InheritanceType;
+import jakarta.persistence.JoinColumn;
+import jakarta.persistence.JoinColumns;
+import jakarta.persistence.ManyToOne;
 
 @Entity
 @Inheritance(strategy = InheritanceType.SINGLE_TABLE)
 @DiscriminatorColumn(name = "task_type", discriminatorType = DiscriminatorType.INTEGER)
 public abstract class TaskInfoEntity {
-    @Id
-    @GeneratedValue
-    private long id;
+    @EmbeddedId
+    private TaskInfoKey taskInfoKey;
 
-    private String jsonPointer;
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumns({
+            @JoinColumn(name = "applicationId", referencedColumnName = "applicationId", insertable = false, updatable = false),
+            @JoinColumn(name = "processInstanceId", referencedColumnName = "instanceId", insertable = false, updatable = false) })
+    private ProcessInstanceEntity processInstance;
 
     public TaskInfoEntity() {
     }
 
-    public TaskInfoEntity(String jsonPointer) {
-        this.jsonPointer = jsonPointer;
+    public TaskInfoEntity(TaskInfoKey taskInfoKey) {
+        this.taskInfoKey = taskInfoKey;
     }
 
     public String jsonPointer() {
-        return jsonPointer;
+        return taskInfoKey.getJsonPointer();
     }
 }

--- a/persistence/jpa/runtime/src/main/java/io/quarkiverse/flow/persistence/jpa/TaskInfoKey.java
+++ b/persistence/jpa/runtime/src/main/java/io/quarkiverse/flow/persistence/jpa/TaskInfoKey.java
@@ -1,0 +1,75 @@
+package io.quarkiverse.flow.persistence.jpa;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+import jakarta.persistence.Column;
+import jakarta.persistence.Embeddable;
+
+import io.serverlessworkflow.impl.TaskContextData;
+import io.serverlessworkflow.impl.WorkflowContextData;
+
+@Embeddable
+public class TaskInfoKey implements Serializable {
+
+    private static final long serialVersionUID = 1L;
+
+    public static TaskInfoKey from(WorkflowContextData workflow, TaskContextData task) {
+        TaskInfoKey key = new TaskInfoKey();
+        key.jsonPointer = task.position().jsonPointer();
+        key.applicationId = workflow.definition().application().id();
+        key.processInstanceId = workflow.instanceData().id();
+        return key;
+    }
+
+    @Column
+    private String jsonPointer;
+
+    @Column
+    private String applicationId;
+
+    @Column
+    private String processInstanceId;
+
+    public String getJsonPointer() {
+        return jsonPointer;
+    }
+
+    public void setJsonPointer(String jsonPointer) {
+        this.jsonPointer = jsonPointer;
+    }
+
+    public String getApplicationId() {
+        return applicationId;
+    }
+
+    public void setApplicationId(String applicationId) {
+        this.applicationId = applicationId;
+    }
+
+    public String getProcessInstanceId() {
+        return processInstanceId;
+    }
+
+    public void setProcessInstanceId(String processInstanceId) {
+        this.processInstanceId = processInstanceId;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(applicationId, jsonPointer, processInstanceId);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
+        if (obj == null)
+            return false;
+        if (getClass() != obj.getClass())
+            return false;
+        TaskInfoKey other = (TaskInfoKey) obj;
+        return Objects.equals(applicationId, other.applicationId) && Objects.equals(jsonPointer, other.jsonPointer)
+                && Objects.equals(processInstanceId, other.processInstanceId);
+    }
+}

